patches/net-cookies-canonical_cookie.cc.patch

```diff
@@ -0,0 +1,60 @@
+diff --git a/net/cookies/canonical_cookie.cc b/net/cookies/canonical_cookie.cc
+index 91611ac4171c19a031044ae6b1459acce246d427..daf61181bcaebef8b53f9e5de43319cc0a26e698 100644
+--- a/net/cookies/canonical_cookie.cc
++++ b/net/cookies/canonical_cookie.cc
+@@ -89,6 +89,11 @@ int PartialCookieOrdering(const CanonicalCookie& a, const CanonicalCookie& b) {
+ 
+ }  // namespace
+ 
++const TimeDelta CanonicalCookie::kMaxClientSideExpiration =
++    TimeDelta::FromSeconds(60*60*24*7);  // 7 days
++const TimeDelta CanonicalCookie::kMaxServerSideExpiration =
++    TimeDelta::FromSeconds(60*60*24*30*6);  // 6 months
++
+ // Keep defaults here in sync with content/public/common/cookie_manager.mojom.
+ CanonicalCookie::CanonicalCookie()
+     : secure_(false),
+@@ -158,7 +163,11 @@ std::string CanonicalCookie::CanonPathWithString(
+ // static
+ Time CanonicalCookie::CanonExpiration(const ParsedCookie& pc,
+                                       const Time& current,
+-                                      const Time& server_time) {
++                                      const Time& server_time,
++                                      const bool is_from_http) {
++  const Time max_expiration = current +
++      (is_from_http ? kMaxServerSideExpiration : kMaxClientSideExpiration);
++
+   // First, try the Max-Age attribute.
+   uint64_t max_age = 0;
+   if (pc.HasMaxAge() &&
+@@ -168,7 +177,7 @@ Time CanonicalCookie::CanonExpiration(const ParsedCookie& pc,
+       sscanf(
+ #endif
+              pc.MaxAge().c_str(), " %" PRIu64, &max_age) == 1) {
+-    return current + TimeDelta::FromSeconds(max_age);
++    return std::min(current + TimeDelta::FromSeconds(max_age), max_expiration);
+   }
+ 
+   // Try the Expires attribute.
+@@ -177,7 +186,7 @@ Time CanonicalCookie::CanonExpiration(const ParsedCookie& pc,
+     base::Time parsed_expiry =
+         cookie_util::ParseCookieExpirationTime(pc.Expires());
+     if (!parsed_expiry.is_null())
+-      return parsed_expiry + (current - server_time);
++      return std::min(parsed_expiry + (current - server_time), max_expiration);
+   }
+ 
+   // Invalid or no expiration, persistent cookie.
+@@ -228,9 +237,11 @@ std::unique_ptr<CanonicalCookie> CanonicalCookie::Create(
+     server_time = options.server_time();
+ 
+   DCHECK(!creation_time.is_null());
++  bool is_from_http = !options.exclude_httponly();
+   Time cookie_expires = CanonicalCookie::CanonExpiration(parsed_cookie,
```

Comment:
we can have an `CanonicalCookie ::BraveCanonExpiration)` calling `CanonicalCookie::CanonExpiration` to get its result and return the min value in `brave/chromium_src/net/cookies/canonical_cookie.cc`.

Also shouldn't we use `ParsedCookie.IsHttpOnly` instead of `CookieOptions.exclude_httponly`?

Reply:
> Also shouldn't we use `ParsedCookie.IsHttpOnly` instead of `CookieOptions.exclude_httponly`?

The naming is a little confusing, but those two functions actually mean something different:
1. `ParsedCookie::IsHttpOnly()` is the `httponly` flag part of the `Set-Cookie` header.
2. `CookieOptions::exclude_httponly()` is a flag set by users of the cookie store to indicate two separate things depending on the context:
   1. whether or not to include `httponly` cookies when doing cookie lookups ("getter path")
   2. whether the cookie creation is happening due to an HTTP header or due to the `document.cookie` DOM API ("setter path")

For this patch, we want to be able to tell the API that the cookie setter is using (i.e. § 2.ii), similar to how `CanonicalCookie::Create()` [uses that flag to reject `httponly` set using `document.cookie`](https://cs.chromium.org/chromium/src/net/cookies/canonical_cookie.cc?l=211-216&rcl=4802de68c80fddacee86ca8427736b006ea1e957).